benchmark/integration/tpcc_benchmark.cpp

```diff
@@ -146,10 +147,90 @@ BENCHMARK_DEFINE_F(TPCCBenchmark, ScaleFactor4WithLogging)(benchmark::State &sta
+
+  // NOLINTNEXTLINE
+  for (auto _ : state) {
+    thread_pool_.Startup();
+    unlink(LOG_FILE_NAME);
+    thread_registry_ = new common::DedicatedThreadRegistry(METRICS_DISABLED);
+    // we need transactions, TPCC database, and GC
+    log_manager_ =
+        new storage::LogManager(LOG_FILE_NAME, num_log_buffers_, log_serialization_interval_, log_persist_interval_,
+                                log_persist_threshold_, &buffer_pool_, common::ManagedPointer(thread_registry_));
+    log_manager_->Start();
+    transaction::TimestampManager timestamp_manager;
+    transaction::DeferredActionManager deferred_action_manager(&timestamp_manager);
+    transaction::TransactionManager txn_manager(&timestamp_manager, &deferred_action_manager, &buffer_pool_, true,
+                                                log_manager_);
+
+    // build the TPCC database
+    auto *const tpcc_db = tpcc_builder.Build();
+
+    // prepare the workers
+    workers.clear();
+    for (int8_t i = 0; i < num_threads_; i++) {
+      workers.emplace_back(tpcc_db);
+    }
+
+    // populate the tables and indexes
+    Loader::PopulateDatabase(&txn_manager, &generator_, tpcc_db, workers);
+    log_manager_->ForceFlush();
+    gc_ = new storage::GarbageCollector(&timestamp_manager, &deferred_action_manager, &txn_manager, DISABLED);
+    gc_thread_ = new storage::GarbageCollectorThread(gc_, gc_period_);
+    Util::RegisterIndexesForGC(&(gc_thread_->GetGarbageCollector()), tpcc_db);
+    std::this_thread::sleep_for(std::chrono::seconds(2));  // Let GC clean up
+
+    // run the TPCC workload to completion, timing the execution
+    uint64_t elapsed_ms;
+    {
+      common::ScopedTimer<std::chrono::milliseconds> timer(&elapsed_ms);
+      for (int8_t i = 0; i < num_threads_; i++) {
+        thread_pool_.SubmitTask([i, tpcc_db, &txn_manager, precomputed_args, &workers] {
+          Workload(i, tpcc_db, &txn_manager, precomputed_args, &workers);
+        });
+      }
+      thread_pool_.WaitUntilAllFinished();
+      log_manager_->ForceFlush();
+    }
+
+    state.SetIterationTime(static_cast<double>(elapsed_ms) / 1000.0);
+
+    // cleanup
+    thread_pool_.Shutdown();
+    log_manager_->PersistAndStop();
+    delete log_manager_;
+    delete gc_thread_;
+    delete thread_registry_;
+    delete gc_;
+    delete tpcc_db;
+  }
+
+  CleanUpVarlensInPrecomputedArgs(&precomputed_args);
+
+  // Count the number of txns processed
+  if (only_count_new_order_) {
+    uint64_t num_new_orders = 0;
+    for (const auto &worker_txns : precomputed_args) {
+      for (const auto &txn : worker_txns) {
+        if (txn.type_ == TransactionType::NewOrder) num_new_orders++;
+      }
+    }
+    state.SetItemsProcessed(state.iterations() * num_new_orders);
+  } else {
+    state.SetItemsProcessed(state.iterations() * num_precomputed_txns_per_worker_ * num_threads_);
+  }
+}
+
+// NOLINTNEXTLINE
+BENCHMARK_DEFINE_F(TPCCBenchmark, ScaleFactor4WithLoggingAndMetrics)(benchmark::State &state) {
```

Comment:
It looks like there're quite some duplicated code in the few functions in this file. Should we abstract them up?